mcxisp/mcxisp.c

```diff
@@ -0,0 +1,616 @@
+
+	return pos;
+}
+
+
+static void cmd_constructFrame(uint8_t *frame, uint8_t type, uint16_t len)
+{
+	size_t pos = 0;
+	uint16_t crc = 0;
+
+	pos += serialize8(frame + pos, FRAME_START);
+	pos += serialize8(frame + pos, type);
+	pos += serialize16(frame + pos, len);
+
+	crc16(&crc, frame, pos);
+	crc16(&crc, frame + FRAME_SIZE, len);
+
+	serialize16(frame + pos, crc);
+}
+
+
+static size_t cmd_flashEraseAll(uint8_t *buff, int memid)
+{
+	size_t pos = FRAME_SIZE; /* Skip frame for now */
+
+	pos += serialize8(buff + pos, 0x01);   /* Tag */
+	pos += serialize8(buff + pos, 0x00);   /* Flags */
+	pos += serialize8(buff + pos, 0x00);   /* Reserved */
+	pos += serialize8(buff + pos, 0x01);   /* Parameter count */
+	pos += serialize32(buff + pos, memid); /* Memory id */
+
+	cmd_constructFrame(buff, kFramingPacketType_Command, pos - FRAME_SIZE);
+
+	return pos;
+}
+
+
+static size_t cmd_flashWriteMemory(uint8_t *buff, uint32_t address, size_t len, int memid)
+{
+	size_t pos = FRAME_SIZE; /* Skip frame for now */
+
+	pos += serialize8(buff + pos, 0x04);     /* Tag */
+	pos += serialize8(buff + pos, 0x01);     /* Flags */
+	pos += serialize8(buff + pos, 0x00);     /* Reserved */
+	pos += serialize8(buff + pos, 0x03);     /* Parameter count */
+	pos += serialize32(buff + pos, address); /* Start address */
+	pos += serialize32(buff + pos, len);     /* Data length */
+	pos += serialize32(buff + pos, memid);   /* Memory id */
+
+	cmd_constructFrame(buff, kFramingPacketType_Command, pos - FRAME_SIZE);
+
+	return pos;
+}
+
+
+static size_t cmd_reset(uint8_t *buff)
+{
+	size_t pos = FRAME_SIZE; /* Skip frame for now */
+
+	pos += serialize8(buff + pos, 0x0b); /* Tag */
+	pos += serialize8(buff + pos, 0x00); /* Flags */
+	pos += serialize8(buff + pos, 0x00); /* Reserved */
+	pos += serialize8(buff + pos, 0x00); /* Parameter count */
+
+	cmd_constructFrame(buff, kFramingPacketType_Command, pos - FRAME_SIZE);
+
+	return pos;
+}
+
+
+static size_t cmd_data(uint8_t *buff, const uint8_t *data, size_t len)
+{
+	memcpy(buff + FRAME_SIZE, data, len);
+	cmd_constructFrame(buff, kFramingPacketType_Data, len);
+	return FRAME_SIZE + len;
+}
+
+
+static int cmd_expectAck(const uint8_t *buff)
+{
+	if ((buff[0] != 0x5a) || (buff[1] != 0xa1)) {
+		fprintf(stderr, "target invalid response\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int cmd_expectGenericResponse(const uint8_t *buff)
+{
+	if ((buff[0] != 0x5a) || (buff[1] != 0xa4)) {
+		fprintf(stderr, "target invalid response\n");
+		return -1;
+	}
+
+	/* Ignore the rest of the packet, not interesting really */
+
+	return 0;
+}
+
+
+static int cmd_expectPingResponse(const uint8_t *buff)
+{
+	if ((buff[0] != 0x5a) || (buff[1] != 0xa7)) {
+		fprintf(stderr, "target invalid response\n");
+		return -1;
+	}
+
+	/* Ignore the rest of the packet, not interesting really */
+
+	return 0;
+}
+
+#if TTY_DEBUG
+static void tty_dump(const uint8_t *buff, size_t len)
+{
+	for (size_t i = 0; i < len; ++i) {
+		printf("%02x", buff[i]);
```

Comment:
How about sth like
```
#define TTY_DEBUG(fmt, ...) if (0) printf(fmt, ##__VA_ARGS__);
```
and use it instead of the `#if`s?

Reply:
Good idea, `for` should be optimized-out in non debug case, so no downsides